File: udp_chat_client.py
#!/usr/bin/env python3
"""
UDP Chat Client for real-time messaging
Implements chat functionality from project requirements
"""

import socket
import threading
import sys

class UDPChatClient:
    def __init__(self, host='localhost', port=9876):
        self.host = host
        self.port = port
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
        self.running = True
        
    def start(self):
        """Start the UDP chat client"""
        print("=== UDP Chat Client ===")
        print(f"Server: {self.host}:{self.port}")
        print("Chat dasturi ishga tushdi!")
        print("Xabarlarni yuboring (chiqish uchun 'exit' yozing):")
        print()
        
        # Start listening for messages in a separate thread
        listener_thread = threading.Thread(target=self.listen_for_messages, daemon=True)
        listener_thread.start()
        
        # Main chat loop
        while self.running:
            try:
                user_input = input("Siz: ").strip()
                
                if user_input.lower() == 'exit':
                    print("Chatdan chiqilmoqda...")
                    self.running = False
                    break
                
                if not user_input:
                    continue
                
                # Send message to server
                self.socket.sendto(user_input.encode('utf-8'), (self.host, self.port))
                
            except KeyboardInterrupt:
                print("\nChatdan chiqilmoqda...")
                self.running = False
                break
            except Exception as e:
                print(f"Xatolik: {e}")
        
        self.socket.close()
    
    def listen_for_messages(self):
        """Listen for incoming messages from server"""
        while self.running:
            try:
                data, addr = self.socket.recvfrom(1024)
                message = data.decode('utf-8')
                print(f"<< {message}")
            except Exception as e:
                if self.running:
                    print(f"Xabar olishda xatolik: {e}")
                break

def main():
    """Main function to start the chat client"""
    client = UDPChatClient()
    try:
        client.start()
    except KeyboardInterrupt:
        print("\nChat client to'xtatilmoqda...")

if __name__ == "__main__":
    main()

File: UDPChatClient.java
package udp;

import java.io.*;
import java.net.*;
import java.util.Scanner;

/**
 * UDP Chat Client for real-time messaging
 * Implements chat functionality from project requirements
 */
class UDPChatClient {
    private static final String SERVER_HOST = "localhost";
    private static final int SERVER_PORT = 9876;
    private static final int BUFFER_SIZE = 1024;
    
    public static void main(String[] args) throws Exception {
        Scanner scanner = new Scanner(System.in);
        DatagramSocket clientSocket = new DatagramSocket();
        InetAddress serverAddress = InetAddress.getByName(SERVER_HOST);
        
        System.out.println("=== UDP Chat Client ===");
        System.out.println("Server: " + SERVER_HOST + ":" + SERVER_PORT);
        System.out.println("Chat dasturi ishga tushdi!");
        System.out.println("Xabarlarni yuboring (chiqish uchun 'exit' yozing):");
        System.out.println();
        
        // Start listening for messages in a separate thread
        Thread listenerThread = new Thread(() -> {
            while (true) {
                try {
                    byte[] receiveData = new byte[BUFFER_SIZE];
                    DatagramPacket receivePacket = new DatagramPacket(receiveData, receiveData.length);
                    clientSocket.receive(receivePacket);
                    
                    String message = new String(receivePacket.getData()).trim();
                    System.out.println("<< " + message);
                    
                } catch (Exception e) {
                    if (!clientSocket.isClosed()) {
                        System.err.println("Xabar olishda xatolik: " + e.getMessage());
                    }
                    break;
                }
            }
        });
        listenerThread.setDaemon(true);
        listenerThread.start();
        
        // Main chat loop
        while (true) {
            try {
                System.out.print("Siz: ");
                String userInput = scanner.nextLine();
                
                if (userInput.equalsIgnoreCase("exit")) {
                    System.out.println("Chatdan chiqilmoqda...");
                    break;
                }
                
                if (userInput.trim().isEmpty()) {
                    continue;
                }
                
                // Send message to server
                byte[] sendData = userInput.getBytes();
                DatagramPacket sendPacket = new DatagramPacket(
                    sendData, sendData.length, serverAddress, SERVER_PORT
                );
                clientSocket.send(sendPacket);
                
            } catch (Exception e) {
                System.err.println("Xatolik: " + e.getMessage());
            }
        }
        
        clientSocket.close();
        scanner.close();
    }
}
